### ISSUE TITLE
ExecutorTest#test_class_serial_is_unaffected always fails

### ISSUE THREAD

**Issue by yahonda** [rails/rails#44475]
### Steps to reproduce

Managed to reproduce https://buildkite.com/rails/rails/builds/84752#ba0366db-f450-475a-a6ca-48202fa35e3e but have no idea why.

```ruby
git clone https://github.com/rails/rails
cd rails/activesupport
bundle install # `bundle update nokogiri --conservative` if you run Ruby 3.1
bin/test test/executor_test.rb -n test_class_serial_is_unaffected
```

### Expected behavior
It should pass.

### Actual behavior
It always fails.

```ruby
$ bin/test test/executor_test.rb -n test_class_serial_is_unaffected
Run options: -n test_class_serial_is_unaffected --seed 37478

# Running:

F

Failure:
ExecutorTest#test_class_serial_is_unaffected [/home/yahonda/rails/activesupport/test/executor_test.rb:241]:
Expected: 2694
  Actual: 2696


bin/test test/executor_test.rb:220



Finished in 0.013746s, 72.7506 runs/s, 72.7506 assertions/s.
1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
$
```

### System configuration
**Rails version**: main branch

**Ruby version**:  It fails against these Rubies.
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-linux]
ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [x86_64-linux]
ruby 2.7.5p203 (2021-11-24 revision f69aeb8314) [x86_64-linux]

**Comment by yahonda** [comment]
Somehow it does NOT reproduce against macOS running M1 chip.

ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [arm64-darwin21]
ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [arm64-darwin21]
ruby 2.7.5p203 (2021-11-24 revision f69aeb8314) [arm64-darwin21]

**Comment by yahonda** [comment]
It has been addressed via e4140140af5832e045bf8196d1bcbe9e5276a8fb
